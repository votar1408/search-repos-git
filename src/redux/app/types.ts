export const LOADING: string = 'LOADING';

export interface AppLoadingAction {
    type: typeof LOADING;
    payload: boolean;
}

export type AppAction = AppLoadingAction;

export interface AppState {
    loading: boolean;
}
