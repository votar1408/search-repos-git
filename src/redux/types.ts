import {AppState} from './app/types';

export interface RootAppState {
    app: AppState;
}

interface RootState extends RootAppState {}

export const selectApp = (state: RootState) => state.app;
