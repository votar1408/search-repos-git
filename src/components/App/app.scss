.main-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    width: 100vw;
    height: calc(100vh - 60px);
    height: 100vh;
    max-width: 1200px;
    margin: 0px auto;
    padding: 60px 0 0 0;
    background: white;
    box-shadow: 0 1px 4px rgba(0, 0, 0, 0.1), -10px 0 20px -10px rgba(0, 0, 0, 0.4),
        10px 0 20px -10px rgba(0, 0, 0, 0.4), 0 0 0 rgba(0, 0, 0, 0.1) inset;
}
