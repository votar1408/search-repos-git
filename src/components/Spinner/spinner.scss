.root {
    display: flex;
    position: absolute;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;
    background: rgba(0, 0, 0, 0.2);
    top: 0;
    left: 0;
}

.root__spinner {
    width: '100px';
}

.spinner-img {
    max-width: '100%';
}
