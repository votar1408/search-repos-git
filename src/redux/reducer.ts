import {combineReducers} from 'redux';

import {appReducer} from './app/reducer';

export const reducer = combineReducers({
    app: appReducer
});
