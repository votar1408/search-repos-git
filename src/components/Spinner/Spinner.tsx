import './spinner.scss';

import React, {FC} from 'react';

import spinner from '../../assets/svg/spinner.svg';

export const Spinner: FC = () => {
    return (
        <div className="root">
            <div className="root__spinner">
                <img className="spinner-img" src={spinner} alt="spinner" />
            </div>
        </div>
    );
};
