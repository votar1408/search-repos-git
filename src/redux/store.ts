/* eslint-disable no-underscore-dangle */
import {
    applyMiddleware,
    compose,
    createStore,
    Middleware
} from 'redux';
import thunk from 'redux-thunk';

import {reducer} from './reducer';

const middleware: [Middleware] = [thunk];

export const store = createStore(reducer, compose(applyMiddleware(...middleware)));
