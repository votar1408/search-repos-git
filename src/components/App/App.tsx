import '../../assets/scss/main.scss';
import '../../assets/scss/nullstyle.scss';
import './app.scss';

import React, {
    FC,
    ReactElement
} from 'react';

import {useSelector} from 'react-redux';

import {selectApp} from '../../redux/types';
import {Header} from '../Header';
import {Spinner} from '../Spinner';

export const App: FC = () => {
    const {loading} = useSelector(selectApp);

    const spinner: ReactElement | null = loading ? <Spinner /> : null;

    return (
        <>
            <Header />
            <main className="main-container">
                Hello
                {spinner}
            </main>
        </>
    );
};
