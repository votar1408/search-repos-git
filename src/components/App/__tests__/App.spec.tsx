import React from 'react';

import {mount} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import {Provider} from 'react-redux';
import {BrowserRouter as Router} from 'react-router-dom';
import createMockStore, {
    MockStoreCreator,
    MockStoreEnhanced
} from 'redux-mock-store';

import {RootAppState} from '../../../redux/types';
import {Spinner} from '../../Spinner';
import {App} from '../App';

describe('App', () => {
    const getInitialState = (value: boolean = false): RootAppState => {
        return {
            app: {
                loading: value,
                repositories: [
                    undefined,
                    [
                        {
                            id: 1,
                            name: 'test1',
                            description: 'some description 1',
                            owner: {
                                login: 'login1',
                                avatar_url: 'some url1'
                            }
                        },
                        {
                            id: 2,
                            name: 'test2',
                            description: 'some description 2',
                            owner: {
                                login: 'login2',
                                avatar_url: 'some url2'
                            }
                        }
                    ]
                ],
                page: 1,
                searchValue: 'test'
            }
        };
    };

    const mockStore: MockStoreCreator = createMockStore();

    let store: MockStoreEnhanced;
    let wrapper: any;

    const getWrapper = (state: RootAppState) => {
        store = mockStore(state);
        return mount(
            <Router>
                <Provider store={store}>
                    <App />
                </Provider>
            </Router>
        );
    };

    beforeEach(() => {
        wrapper = getWrapper(getInitialState());
    });

    it('should render correctly', () => {
        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });

    it('renders without crashing', () => {
        expect(wrapper.find(App).length).toEqual(1);
    });

    it('should contains 1 <main/>', () => {
        expect(wrapper.find('main').length).toEqual(1);
    });

    it('not should contains <Spinner/> component if param loading equal false', () => {
        expect(wrapper.find(Spinner).length).toEqual(0);
    });

    it('should contains 1 <Spinner/> if param loading equal true', () => {
        wrapper = getWrapper(getInitialState(true));
        expect(wrapper.find(Spinner).length).toEqual(1);
    });
});
