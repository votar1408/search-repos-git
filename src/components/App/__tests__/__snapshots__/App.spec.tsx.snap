// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <App>
        <Header>
          <header
            className="header-root"
          >
            <div>
              <span>
                Search of Repositories in GitHub
              </span>
            </div>
          </header>
        </Header>
        <main
          className="main-container"
        >
          <RoutePage>
            <Switch>
              <Route
                component={[Function]}
                computedMatch={
                  Object {
                    "isExact": true,
                    "params": Object {},
                    "path": "/",
                    "url": "/",
                  }
                }
                exact={true}
                location={
                  Object {
                    "hash": "",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  }
                }
                path="/"
              >
                <PageHome
                  history={
                    Object {
                      "action": "POP",
                      "block": [Function],
                      "createHref": [Function],
                      "go": [Function],
                      "goBack": [Function],
                      "goForward": [Function],
                      "length": 1,
                      "listen": [Function],
                      "location": Object {
                        "hash": "",
                        "pathname": "/",
                        "search": "",
                        "state": undefined,
                      },
                      "push": [Function],
                      "replace": [Function],
                    }
                  }
                  location={
                    Object {
                      "hash": "",
                      "pathname": "/",
                      "search": "",
                      "state": undefined,
                    }
                  }
                  match={
                    Object {
                      "isExact": true,
                      "params": Object {},
                      "path": "/",
                      "url": "/",
                    }
                  }
                >
                  <SearchRepositories>
                    <form
                      className="search"
                      onSubmit={[Function]}
                    >
                      <input
                        className="search__term"
                        name="search-repo-input"
                        onChange={[Function]}
                        placeholder="Input the name of the repository"
                        type="text"
                        value=""
                      />
                      <button
                        className="search__button"
                        name="submit-button"
                        type="submit"
                      >
                        <i
                          className="fa fa-search"
                        />
                      </button>
                    </form>
                  </SearchRepositories>
                  <RepositoriesList>
                    <ul
                      className="repositories-container"
                    >
                      <Repository
                        item={
                          Object {
                            "description": "some description 1",
                            "id": 1,
                            "name": "test1",
                            "owner": Object {
                              "avatar_url": "some url1",
                              "login": "login1",
                            },
                          }
                        }
                        key="1"
                      >
                        <li
                          className="repository"
                        >
                          <div
                            className="repository__block"
                            data-id="repository-block"
                          >
                            <span
                              className="repository__block_title"
                            >
                              Name:
                            </span>
                            <span
                              className="repository__block_description"
                            >
                              test1
                            </span>
                          </div>
                          <div
                            className="repository__block"
                            data-id="repository-block"
                          >
                            <span
                              className="repository__block_title"
                            >
                              Description:
                            </span>
                            <span
                              className="repository__block_description"
                            >
                              some description 1
                            </span>
                          </div>
                          <div
                            className="repository__block"
                            data-id="repository-block"
                          >
                            <span
                              className="repository__block_title"
                            >
                              Owner:
                            </span>
                            <img
                              alt="logo"
                              className="repository__block_logo"
                              src="some url1"
                            />
                            <span
                              className="repository__block_description"
                            >
                              login1
                            </span>
                          </div>
                        </li>
                      </Repository>
                      <Repository
                        item={
                          Object {
                            "description": "some description 2",
                            "id": 2,
                            "name": "test2",
                            "owner": Object {
                              "avatar_url": "some url2",
                              "login": "login2",
                            },
                          }
                        }
                        key="2"
                      >
                        <li
                          className="repository"
                        >
                          <div
                            className="repository__block"
                            data-id="repository-block"
                          >
                            <span
                              className="repository__block_title"
                            >
                              Name:
                            </span>
                            <span
                              className="repository__block_description"
                            >
                              test2
                            </span>
                          </div>
                          <div
                            className="repository__block"
                            data-id="repository-block"
                          >
                            <span
                              className="repository__block_title"
                            >
                              Description:
                            </span>
                            <span
                              className="repository__block_description"
                            >
                              some description 2
                            </span>
                          </div>
                          <div
                            className="repository__block"
                            data-id="repository-block"
                          >
                            <span
                              className="repository__block_title"
                            >
                              Owner:
                            </span>
                            <img
                              alt="logo"
                              className="repository__block_logo"
                              src="some url2"
                            />
                            <span
                              className="repository__block_description"
                            >
                              login2
                            </span>
                          </div>
                        </li>
                      </Repository>
                    </ul>
                  </RepositoriesList>
                  <Pagination>
                    <div
                      className="pagination"
                    >
                      <button
                        className="pagination__button"
                        data-id="left-button"
                        disabled={true}
                        onClick={[Function]}
                        type="submit"
                      >
                        <i
                          className="fa fa-angle-left"
                        />
                      </button>
                      <button
                        className="pagination__button"
                        data-id="right-button"
                        disabled={false}
                        onClick={[Function]}
                        type="submit"
                      >
                        <i
                          className="fa fa-angle-right"
                        />
                      </button>
                    </div>
                  </Pagination>
                </PageHome>
              </Route>
            </Switch>
          </RoutePage>
        </main>
      </App>
    </Provider>
  </Router>
</BrowserRouter>
`;
